README
NAME
    App::ShellCompleter::cpanm - Shell completion for cpanm

VERSION
    This document describes version 0.07 of App::ShellCompleter::cpanm (from
    Perl distribution App-ShellCompleter-cpanm), released on 2015-01-07.

SYNOPSIS
    See _cpanm included in this distribution.

SEE ALSO
    Bash::Completion::Plugins::cpanm, which focuses on completing module
    name remotely using MetaCPAN API. This module, on the other hand,
    focuses on completing "cpanm" command-line options.

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/App-ShellCompleter-cpanm>.

SOURCE
    Source repository is at
    <https://github.com/perlancar/perl-App-BashCompleter-cpanm>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=App-ShellCompleter-cp
    anm>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    perlancar <perlancar@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2015 by perlancar@cpan.org.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for App-ShellCompleter-cpanm

0.08    2015-01-08 (PERLANCAR)

	- [dzil] Update prereqs.

	- Disable the use of XPAN::Query (completing available CPAN
	  modules) for now due to slow initial response time.


0.07    2015-01-07 (PERLANCAR)

	- [opt] Don't bother to complete module names if word doesn't look
	  like module name.


0.06    2014-12-31 (PERLANCAR)

	- Tweak logging, fix completion of CPAN modules.


0.05    2014-12-31 (PERLANCAR)

	- Fix/update code (use combine_answers).


0.04    2014-12-29 (PERLANCAR)

        - No functional changes.

        - Remove FAQ item because we now have a workaround in Complete::Bash.


0.03    2014-12-14 (PERLANCAR)

        - Rename module/dist from App-BashCompleter-cpanm to
          App-ShellCompleter-cpanm.

        - Mention shcompgen instead of the obsolete bash-completion-prog.

        - Complete module names on CPAN using XPAN::Query.

        - Add some logging for debugging.


0.02    2014-11-10 (PERLANCAR)

        - No functional changes.

        - Mention Bash::Completion::Plugins::cpanm.


0.01    2014-11-10 (PERLANCAR)

        - First release.


lib/App/ShellCompleter/cpanm.pm
package App::ShellCompleter::cpanm;

our $DATE = '2015-01-08'; # DATE
our $VERSION = '0.08'; # VERSION

1;
# ABSTRACT: Shell completion for cpanm

__END__

=pod

=encoding UTF-8

=head1 NAME

App::ShellCompleter::cpanm - Shell completion for cpanm

=head1 VERSION

This document describes version 0.08 of App::ShellCompleter::cpanm (from Perl distribution App-ShellCompleter-cpanm), released on 2015-01-08.

=head1 SYNOPSIS

See L<_cpanm> included in this distribution.

=head1 SEE ALSO

L<Bash::Completion::Plugins::cpanm>, which focuses on completing module name
remotely using MetaCPAN API. This module, on the other hand, focuses on
completing C<cpanm> command-line options.

=head1 HOMEPAGE

Please visit the project's homepage at L<https://metacpan.org/release/App-ShellCompleter-cpanm>.

=head1 SOURCE

Source repository is at L<https://github.com/perlancar/perl-App-BashCompleter-cpanm>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://rt.cpan.org/Public/Dist/Display.html?Name=App-ShellCompleter-cpanm>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

perlancar <perlancar@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by perlancar@cpan.org.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.030.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Shell completion for cpanm",
  "AUTHOR" => "perlancar <perlancar\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Perl::osnames" => "0.09"
  },
  "DISTNAME" => "App-ShellCompleter-cpanm",
  "EXE_FILES" => [
    "bin/_cpanm"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "App::ShellCompleter::cpanm",
  "PREREQ_PM" => {
    "Complete::Module" => "0.12",
    "Complete::Util" => 0,
    "Getopt::Long::Complete" => 0,
    "Log::Any" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.08",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Complete::Module" => "0.12",
  "Complete::Util" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Getopt::Long::Complete" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Log::Any" => 0,
  "Perl::osnames" => "0.09",
  "Test::More" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);




# modify generated Makefile to generate shell completion scripts. this piece
# is generated by Dist::Zilla::Plugin::GenShellCompletion version 0.07.
GEN_SHELL_COMPLETION:
{
    use Perl::osnames 0.09 qw(is_posix);
    last unless is_posix();

    print "Modifying Makefile to generate shell completion on install\n";
    open my($fh), "<", "Makefile" or die "Can't open generated Makefile: $!";
    my $content = do { local $/; ~~<$fh> };

    $content =~ s/^(install :: pure_install doc_install)/$1 comp_install/m
        or die "Can't find pattern in Makefile (1)";

    $content =~ s/^(uninstall :: .+)/$1 comp_uninstall/m
        or die "Can't find pattern in Makefile (2)";

    $content .= qq|\ncomp_install :\n\t| .
        q|$(PERLRUN) -E'if(eval { require App::shcompgen; 1 }) { system "shcompgen", "--verbose", "generate", "--replace", @ARGV }' -- $(EXE_FILES)| .
        qq|\n\n|;

    $content .= qq|\ncomp_uninstall :\n\t| .
        q|$(PERLRUN) -E'if(eval { require App::shcompgen; 1 }) { system "shcompgen", "--verbose", "remove", @ARGV }' -- $(EXE_FILES)| .
        qq|\n\n|;

    open $fh, ">", "Makefile" or die "Can't write modified Makefile: $!";
    print $fh $content;
}
